File: Data Structures & Algorithms/house-robber/submission-2.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    rob(nums) {
        if(nums.length === 1) return nums[0];
        const notebook = {};
        function getChoriKaMaal(houseNumber){
            if(houseNumber in notebook) return notebook[houseNumber];
            if(houseNumber >= nums.length) return 0;

            notebook[houseNumber] =  Math.max(nums[houseNumber] + getChoriKaMaal(houseNumber+2), getChoriKaMaal(houseNumber+1));
            return notebook[houseNumber];
        }

        return Math.max(getChoriKaMaal(0),getChoriKaMaal(1));
    }
}

File: Data Structures & Algorithms/house-robber-ii/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:

        if len(nums) == 1:
            return nums[0]
        def robHouse(nums):
            cache = {}
            def house(i):
                if i in cache:
                    return cache[i]
                if i >= len(nums):
                    return 0
                cache[i] =  max(house(i+2) + nums[i], house(i+1))
                return cache[i]
            return house(0)

        return max(robHouse(nums[1:]),robHouse(nums[:-1]))
        